From 9bdf4fa54b3a3535f6a5e7535ca20dd6065cef9d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Julian=20R=C3=BCth?= <julian.rueth@fsfe.org>
Date: Sat, 8 Oct 2022 17:28:04 +0200
Subject: [PATCH] Adapt to namespace changes in SageMath 9.7

---
 setup.py | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index 28d9ccc..ee888ab 100755
--- a/setup.py
+++ b/setup.py
@@ -122,6 +122,16 @@ class build_py(_build_py):
                 if mod != 'generalized_multiple_zeta_values']
         return modules
 
+# Work around changes in SageMath 9.7, see https://trac.sagemath.org/wiki/ReleaseTours/sage-9.7#Packagessagesage.rings...arenownamespaces
+try:
+    from sage.misc.package_dir import cython_namespace_package_support
+except (ModuleNotFoundError, ImportError):
+    from contextlib import nullcontext
+    cython_namespace_package_support = nullcontext
+
+with cython_namespace_package_support():
+    extensions = cythonize(extensions)
+
 setup(name='surface_dynamics',
       version=version,
       description="Dynamics on surfaces",
@@ -143,7 +153,7 @@ setup(name='surface_dynamics',
           'surface_dynamics/databases': ['cylinder_diagrams/cyl_diags*', 'generalized_permutation_twins/twins*'],
           'surface_dynamics/flat_surfaces/origamis': ['origamis.db'],
           },
-    ext_modules=cythonize(extensions),
+    ext_modules=extensions,
     classifiers=[
       'Development Status :: 4 - Beta',
       'Intended Audience :: Science/Research',
-- 
2.37.3

